### PR TITLE
[Gekidou] Sidebar Categories FlatList Optimisations

### PR DESCRIPTION
**Pull request by shazm**
#### Summary

Adds a few performance optimisations to the `FlatList` rendering the category / channel list in the sidebar.

#### Ticket Link

Fixes https://mattermost.atlassian.net/browse/MM-42333

#### Checklist

- [x] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: 

#### Release Note

```release-note

```

### CODE REVIEW

File: app/components/channel_list/categories/categories.tsx [hunk 1]
```diff
@@ -44,6 +48,13 @@ const Categories = (props: Props) => {
             style={styles.flex}
             showsHorizontalScrollIndicator={false}
             showsVerticalScrollIndicator={false}
+            keyExtractor={extractKey}
```
Comment (avinashlng1080) [inline app/components/channel_list/categories/categories.tsx:51]: In terms of time or memory usage, what is the gain here ?

File: app/constants/list.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+const VISIBILITY_CONFIG_DEFAULTS = {
+    itemVisiblePercentThreshold: 100,
```
Comment (avinashlng1080) [inline app/constants/list.ts:5]: I'm guessing that to count a category as viewable, the whole text should be clearly visible.

File: app/components/channel_list/categories/body/category_body.tsx [hunk 3]
```diff
@@ -40,6 +44,16 @@ const CategoryBody = ({category, categoryChannels, channels, myChannels}: Props)
         <FlatList
             data={data}
             renderItem={ChannelItem}
+            keyExtractor={extractKey}
+            removeClippedSubviews={true}
+            initialNumToRender={20}
+            windowSize={15}
+            updateCellsBatchingPeriod={10}
+            maxToRenderPerBatch={10}
+            getItemLayout={(d, index) => (
```
Comment (avinashlng1080) [inline app/components/channel_list/categories/body/category_body.tsx:53]: extract to a callback outside the `CategoryBody` component

**Review by avinashlng1080 — CHANGES_REQUESTED** [review]

**Review by larkox — APPROVED** [review]
Apart of the other comments, LGTM.

File: app/components/channel_list/categories/categories.tsx (hunk 1 above)
Comment (enahum) [reply app/components/channel_list/categories/categories.tsx:51]: what is your question about @avinashlng1080 ?

> Used to extract a unique key for a given item at the specified index. Key is used for caching and as the react key to track item re-ordering. The default extractor checks item.key, then item.id, and then falls back to using the index, like React does.

File: app/components/channel_list/categories/body/category_body.tsx [hunk 4]
```diff
@@ -40,6 +44,16 @@ const CategoryBody = ({category, categoryChannels, channels, myChannels}: Props)
         <FlatList
             data={data}
             renderItem={ChannelItem}
+            keyExtractor={extractKey}
+            removeClippedSubviews={true}
+            initialNumToRender={20}
+            windowSize={15}
+            updateCellsBatchingPeriod={10}
+            maxToRenderPerBatch={10}
```
Comment (enahum) [inline app/components/channel_list/categories/body/category_body.tsx:52]: default is already 10

File: app/components/channel_list/categories/categories.tsx [hunk 5]
```diff
@@ -44,6 +48,13 @@ const Categories = (props: Props) => {
...
+            removeClippedSubviews={true}
+            initialNumToRender={5}
+            windowSize={15}
+            updateCellsBatchingPeriod={10}
+            maxToRenderPerBatch={5}
```
Comment (enahum) [inline app/components/channel_list/categories/categories.tsx:56]: why make it smaller than the default? what is the gain here?

File: app/components/channel_list/categories/body/category_body.tsx [hunk 6]
```diff
@@ -40,6 +44,16 @@ const CategoryBody = ({category, categoryChannels, channels, myChannels}: Props)
...
+            getItemLayout={(d, index) => (
+                {length: 40, offset: 40 * index, index}
+            )}
+            viewabilityConfig={List.VISIBILITY_CONFIG_DEFAULTS}
```
Comment (enahum) [inline app/components/channel_list/categories/body/category_body.tsx:56]: why do we care about the visibility of items in this list?

File: app/components/channel_list/categories/categories.tsx [hunk 7]
```diff
@@ -44,6 +48,13 @@ const Categories = (props: Props) => {
...
+            viewabilityConfig={List.VISIBILITY_CONFIG_DEFAULTS}
```
Comment (enahum) [inline app/components/channel_list/categories/categories.tsx:57]: why do we care about the visibility of items in this list?

File: app/components/channel_list/categories/categories.tsx (hunk 5 above)
Comment (shazm) [reply app/components/channel_list/categories/categories.tsx:56]: Broadly;
1. Render a lower count of items, but run the render more often
2. Render a higher count of items, but run the less often
 

Both have tradeoffs, but option 2 means the user *will* see an empty list while the batching catches up to their scroll position. Option 1 uses more processing power (offset by lower amounts to render), but means the UI is updating more often and is more likely to keep up with a users scroll.

https://reactnative.dev/docs/optimizing-flatlist-configuration#updatecellsbatchingperiod

File: app/components/channel_list/categories/body/category_body.tsx (hunk 6 above)
Comment (shazm) [reply app/components/channel_list/categories/body/category_body.tsx:56]: Whoops - for this PR we don't; it's more optimisations for when collapsing is happening.

File: app/components/channel_list/categories/categories.tsx (hunk 7 above)
Comment (shazm) [reply app/components/channel_list/categories/categories.tsx:57]: Same as above - will remove and add later when needed.

File: app/components/channel_list/categories/categories.tsx [hunk 8]
```diff
@@ -44,6 +48,13 @@ const Categories = (props: Props) => {
             style={styles.flex}
             showsHorizontalScrollIndicator={false}
             showsVerticalScrollIndicator={false}
+            keyExtractor={extractKey}
+            removeClippedSubviews={true}
+            initialNumToRender={5}
+            windowSize={15}
```
[the remark this answers is not in the document]
Comment (shazm) [reply app/components/channel_list/categories/categories.tsx:54]: Offset the memory usage with the changes from `updateCellsBatchingPeriod` + `maxToRenderPerBatch`.

File: app/components/channel_list/categories/categories.tsx (hunk 5 above)
Comment (enahum) [reply app/components/channel_list/categories/categories.tsx:54]: The idea is that we give this values a try and then if needed we can start adjusting if it feels that is rendering too much or scrolling generate blank areas.

**Review by enahum — APPROVED** [review]

**Comment by enahum** [comment]
@avinashlng1080 there are two approvals so I'll be merging.. if you have any more concerns let's talk about them in the channel.

File: app/components/channel_list/categories/categories.tsx (hunk 1 above)
Comment (avinashlng1080) [reply app/components/channel_list/categories/categories.tsx:49]: I was referring to the whole optimizations not this 'extractKey'.  Just wanted to know by how much we are winning with this new implementation.

**Pull request merged**